### PR TITLE
Prevent ghost-pings in pypi command

### PR DESCRIPTION
**Pull request by TizzySaurus**
Solves #1624
Solves #1695

Will no longer delete the invoking message of `!pypi` command when the given package is invalid if the message contains user/role mentions. Always allows deletion of error message through reactions.

`utils.messages.wait_for_deletion()` now also [clears the added reaction(s) once timeout expires.](https://github.com/python-discord/bot/commit/df3d011bd6490eadc2b611c8d21062b14ebd18c4)

### CODE REVIEW

File: bot/exts/info/pypi.py [hunk 1]
```diff
@@ -67,8 +67,14 @@ async def get_package_info(self, ctx: Context, package: str) -> None:
                     log.trace(f"Error when fetching PyPi package: {response.status}.")
 
         if error:
-            await ctx.send(embed=embed, delete_after=INVALID_INPUT_DELETE_DELAY)
-            await ctx.message.delete(delay=INVALID_INPUT_DELETE_DELAY)
+            error_message = await ctx.send(embed=embed)
+            await wait_for_deletion(error_message, (ctx.author.id,), timeout=INVALID_INPUT_DELETE_DELAY)
+
+            # If won't ghost-ping when deleting message
```
Comment (Bluenix2) [inline bot/exts/info/pypi.py:73]: Please look over this again haha

Comment (TizzySaurus) [reply bot/exts/info/pypi.py]: Was referring to the fact I didn't need to import `bot.constants.Emojis` anymore. Import was removed

**Review by NIRDERIi — APPROVED** [review]
The code changed should work just fine...

File: bot/exts/info/pypi.py [hunk 2]
```diff
@@ -1,14 +1,16 @@
+import contextlib
```
Comment (Bluenix2) [inline bot/exts/info/pypi.py:1]: Could be good to be consistent, everywhere else it's `from contextlib import suppress`.

File: bot/exts/info/pypi.py [hunk 3]
```diff
@@ -67,8 +69,14 @@ async def get_package_info(self, ctx: Context, package: str) -> None:
                     log.trace(f"Error when fetching PyPi package: {response.status}.")
 
         if error:
-            await ctx.send(embed=embed, delete_after=INVALID_INPUT_DELETE_DELAY)
-            await ctx.message.delete(delay=INVALID_INPUT_DELETE_DELAY)
+            error_message = await ctx.send(embed=embed)
+            await wait_for_deletion(error_message, (ctx.author.id,), timeout=INVALID_INPUT_DELETE_DELAY)
+
+            # If won't ghost-ping when deleting message
```
Comment (Bluenix2) [inline bot/exts/info/pypi.py:75]: This comment is still very weirdly worded.

File: bot/exts/info/pypi.py [hunk 4]
```diff
@@ -67,8 +69,14 @@ async def get_package_info(self, ctx: Context, package: str) -> None:
                     log.trace(f"Error when fetching PyPi package: {response.status}.")
 
         if error:
-            await ctx.send(embed=embed, delete_after=INVALID_INPUT_DELETE_DELAY)
-            await ctx.message.delete(delay=INVALID_INPUT_DELETE_DELAY)
+            error_message = await ctx.send(embed=embed)
+            await wait_for_deletion(error_message, (ctx.author.id,), timeout=INVALID_INPUT_DELETE_DELAY)
```
Comment (Bluenix2) [inline bot/exts/info/pypi.py:73]: I want to know how other feels about this:
```suggestion
            await wait_for_deletion(await ctx.send(embed=embed), (ctx.author.id,), timeout=INVALID_INPUT_DELETE_DELAY)
```
It could degrade readability and make less sense.

File: bot/utils/messages.py [hunk 5]
```diff
@@ -95,8 +94,11 @@ async def wait_for_deletion(
         allow_mods=allow_mods,
     )
 
-    with contextlib.suppress(asyncio.TimeoutError):
+    try:
         await bot.instance.wait_for('reaction_add', check=check, timeout=timeout)
+    except asyncio.TimeoutError:
+        await message.clear_reactions()
```
Comment (Bluenix2) [inline bot/utils/messages.py:100]: Perhaps we could adjust the docstring for this change?

File: bot/exts/info/pypi.py (hunk 3 above)
Comment (TizzySaurus) [reply bot/exts/info/pypi.py:75]: Forgot to reword it when pasting over. Will do that

File: bot/exts/info/pypi.py (hunk 2 above)
Comment (TizzySaurus) [reply bot/exts/info/pypi.py:1]: Wasn't sure if it would change the import orders (I don't really understand how the linting works for imports)

File: bot/exts/info/pypi.py (hunk 4 above)
Comment (TizzySaurus) [reply bot/exts/info/pypi.py:73]: I considered that but opted not to in the end. If others think it's a good idea though I can commit this.

File: bot/utils/messages.py (hunk 5 above)
Comment (TizzySaurus) [reply bot/utils/messages.py:101]: Good idea. [Committed](https://github.com/python-discord/bot/pull/1696/commits/aa77ce65dff915611ce5bc56d77f41b5803a5e65)

File: bot/exts/info/pypi.py (hunk 2 above)
Comment (TizzySaurus) [reply bot/exts/info/pypi.py]: [Committed](https://github.com/python-discord/bot/pull/1696/commits/d52bf9d921465320ba8f70fb36933aed8340480d) with no issues

File: bot/utils/messages.py (hunk 5 above)
Comment (TizzySaurus) [reply bot/utils/messages.py:102]: And [re-committed.](https://github.com/python-discord/bot/pull/1696/commits/1abd1c230aedea6c0e8e680253bd7690cc27fd4c)

**Review by Bluenix2 — APPROVED** [review]
Same as last time, can't test it but it looks good! Nice work

**Comment by TizzySaurus** [comment]
> Same as last time, can't test it but it looks good! Nice work

Thanks! Appreciate your help today :+1:

**Review by vcokltfre — APPROVED** [review]
LGTM!

File: bot/exts/info/doc/_cog.py [hunk 6]
```diff
@@ -342,29 +340,15 @@ async def get_command(self, ctx: commands.Context, *, symbol_name: Optional[str]
...
-                        await error_message.clear_reaction(DELETE_ERROR_MESSAGE_REACTION)
-
-                else:
-                    await wait_for_deletion(error_message, (ctx.author.id,), timeout=NOT_FOUND_DELETE_DELAY)
+                # Make sure that we won't cause a ghost-ping by deleting the message
+                if not (ctx.message.mentions or ctx.message.role_mentions):
                     with suppress(discord.NotFound):
                         await ctx.message.delete()
-                    with suppress(discord.NotFound):
-                        await error_message.delete()
```
Comment (kosayoda) [inline bot/exts/info/doc/_cog.py]: Is this change intentional? Prior, both the invoking message and the error embed would be deleted, but now the error embed remains. I noticed that this is a duplicate commit of the other ghost ping PR, but the situation is still similar to the pypi command.

**Review by kosayoda — CHANGES_REQUESTED** [review]

**Review by jchristgit — APPROVED** [review]
Straightforward and reasonable. Very nice. 👍

File: bot/exts/info/doc/_cog.py (hunk 6 above)
Comment (TizzySaurus) [reply bot/exts/info/doc/_cog.py]: Yeah, I changed this deliberately assuming we'd want consistent behaviour with the docs command. If you think this should be reverted to the old behaviour then I'll do so, but my thoughts was that this should be updated to be the same as docs.

Comment (TizzySaurus) [reply bot/exts/info/doc/_cog.py]: For now I'll leave this conversation unsolved and wait for your response @kosayoda.

Comment (kosayoda) [reply bot/exts/info/doc/_cog.py]: I am asking this question for both this and the docs PR. `wait_for_deletion` does not delete the embed after the timeout. Before the docs PR, the error embed would be deleted after the timeout (code being the removed lines), which is no longer the case now.

Comment (TizzySaurus) [reply bot/exts/info/doc/_cog.py]: Right, I see. I personally prefer keeping the embed and just removing the reactions, but if we want to revert to old system we can do.

Comment (kosayoda) [reply bot/exts/info/doc/_cog.py]: For these embeds specifically, the idea is to not have multiple error embeds lying around, since they have lose meaning right after being shown. This is in line with the short configured delay on the reaction check.

Comment (Bluenix2) [reply bot/exts/info/doc/_cog.py]: I think we should add the embed delete inside the same `suppress()`, that way the embed only gets deleted if the bot is able to delete the invoking message.

See this diff:
```diff
                # Make sure that we won't cause a ghost-ping by deleting the message
                if not (ctx.message.mentions or ctx.message.role_mentions):
                    with suppress(discord.NotFound):
                        await ctx.message.delete()
+                       await error_message.delete()
```

If the invoking message pings someone, and they did not react with a trashcan, then both messages will stay up.
If the invoking messaes does not ping someone both messages will be deleted regardless of reacting with a trashcan in time.

Comment (TizzySaurus) [reply bot/exts/info/doc/_cog.py]: That sounds good to me. I'll do this if @kosayoda thinks that's good too.

Comment (TizzySaurus) [reply bot/exts/info/doc/_cog.py]: Kosayoda has approved on discord so will commit this :+1:

**Review by kosayoda — APPROVED** [review]
Works well!
